File: Sources/macOS/Packages/Yage/Sources/Yage/Capabilities/AnimationsScheduler.swift
import Foundation
import Schwifty

public class AnimationsScheduler: Capability {
    public override func install(on subject: Entity) {
        super.install(on: subject)
        DispatchQueue.main.async { [weak self] in
            self?.scheduleAnimationAfterRandomInterval()
        }
    }

    public func animateNow() {
        guard let animation = subject?.animationsProvider?.randomAnimation() else { return }
        let loops = animation.requiredLoops ?? Int.random(in: 1 ... 2)
        schedule(animation, times: loops, after: 0)
        Logger.log(tag, "Immediate animation requested", animation.description, "x\(loops)")
    }

    private func scheduleAnimationAfterRandomInterval() {
        guard let animation = subject?.animationsProvider?.randomAnimation() else { return }
        let delay = TimeInterval.random(in: 10...30)
        let loops = animation.requiredLoops ?? 1
        schedule(animation, times: loops, after: delay)
        Logger.log(tag, "Scheduled", animation.description, "x\(loops) in \(delay)\"")
    }

    public func schedule(_ animation: EntityAnimation, times: Int, after delay: TimeInterval) {
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let self else { return }
            guard self.isEnabled && self.subject?.isAlive == true else { return }
            self.load(animation, times: times)
            self.scheduleAnimationAfterRandomInterval()
        }
    }

    public func load(_ animation: EntityAnimation, times: Int) {
        guard case .move = subject?.state else { return }
        Logger.log(tag, "Loading", animation.description)
        subject?.set(state: .action(action: animation, loops: times))
    }
}

public extension Entity {
    var animationsScheduler: AnimationsScheduler? {
        capability(for: AnimationsScheduler.self)
    }
}

File: Sources/windows/yage/capabilities/animations_scheduler.py
from datetime import datetime, timedelta
import random
from yage.models.animations import EntityAnimation
from yage.models.capability import Capability
from yage.models.entity_state import EntityState
from yage.utils.logger import Logger


class AnimationsScheduler(Capability):
    def __init__(self, subject):
        super().__init__(subject)
        subject.animations_scheduler = self
        self._scheduled_animation = None

    def animate_now(self):
        animation = self._random_animation()
        if not animation:
            return
        loops = animation.required_loops or 1
        self._schedule(animation, loops, 0)
        Logger.log(self.tag, "Immediate animation requested", animation, f"x{loops}")

    def load_animation(self, animation: EntityAnimation, times: int):
        if not self.is_enabled:
            return
        if not self.subject.is_alive:
            return
        if self.subject.state != EntityState.MOVE:
            return
        self.subject.set_animation(animation, times)

    def do_update(self, collisions, time):
        if self._scheduled_animation is not None:
            self._apply_scheduled_sprite_if_possible()
        else:
            self._schedule_animation_after_random_interval()

    def _schedule_animation_after_random_interval(self):
        animation = self._random_animation()
        if not animation:
            return
        delay = random.uniform(10, 30)
        loops = animation.required_loops or 1
        self._schedule(animation, loops, delay)
        Logger.log(self.tag, "Scheduled", animation, f"x{loops} in {delay}")

    def _random_animation(self):
        try:
            return self.subject.animations_provider.random_animation()
        except TypeError:
            Logger.log(self.tag, "Could not load any animation...")
            return None

    def _schedule(self, animation: EntityAnimation, times: int, delay: float):
        next_date = datetime.now() + timedelta(seconds=delay)
        self._scheduled_animation = (animation, times, next_date)

    def _apply_scheduled_sprite_if_possible(self):
        animation, times, date = self._scheduled_animation
        if date > datetime.now():
            return
        self._scheduled_animation = None
        self.load_animation(animation, times)
